src/win/async.c

```diff
@@ -23,77 +23,70 @@
 
 #include "uv.h"
 #include "internal.h"
-#include "atomicops-inl.h"
 #include "handle-inl.h"
 #include "req-inl.h"
 
 
 void uv_async_endgame(uv_loop_t* loop, uv_async_t* handle) {
-  if (handle->flags & UV__HANDLE_CLOSING &&
-      !handle->async_sent) {
-    assert(!(handle->flags & UV_HANDLE_CLOSED));
-    uv__handle_close(handle);
-  }
+  assert(handle->flags & UV__HANDLE_CLOSING);
+  assert(!(handle->flags & UV_HANDLE_CLOSED));
+  uv__handle_close(handle);
 }
 
 
 int uv_async_init(uv_loop_t* loop, uv_async_t* handle, uv_async_cb async_cb) {
-  uv_req_t* req;
-
   uv__handle_init(loop, (uv_handle_t*) handle, UV_ASYNC);
   handle->async_sent = 0;
   handle->async_cb = async_cb;
 
-  req = &handle->async_req;
-  uv_req_init(loop, req);
-  req->type = UV_WAKEUP;
-  req->data = handle;
-
+  QUEUE_INSERT_TAIL(&loop->async_handles, &handle->queue);
   uv__handle_start(handle);
 
   return 0;
 }
 
 
 void uv_async_close(uv_loop_t* loop, uv_async_t* handle) {
-  if (!((uv_async_t*)handle)->async_sent) {
-    uv_want_endgame(loop, (uv_handle_t*) handle);
-  }
-
+  QUEUE_REMOVE(&handle->queue);
+  uv_want_endgame(loop, (uv_handle_t*) handle);
   uv__handle_closing(handle);
 }
 
 
 int uv_async_send(uv_async_t* handle) {
-  uv_loop_t* loop = handle->loop;
-
-  if (handle->type != UV_ASYNC) {
-    /* Can't set errno because that's not thread-safe. */
-    return -1;
-  }
-
-  /* The user should make sure never to call uv_async_send to a closing */
-  /* or closed handle. */
-  assert(!(handle->flags & UV__HANDLE_CLOSING));
+  /* First do a cheap read. */
+  if (handle->async_sent != 0)
+    return 0;
 
-  if (!uv__atomic_exchange_set(&handle->async_sent)) {
-    POST_COMPLETION_FOR_REQ(loop, &handle->async_req);
+  if (InterlockedExchange(&handle->async_sent, 1) == 0) {
+    uv_loop_t* loop = handle->loop;
```

Comment:
Ah, I see.  You can't really rely on the compiler actually allocating a stack slot or a register, though.  It might simply decide to read handle->loop twice.